services/src/main/java/org/keycloak/protocol/oidc/grants/ciba/endpoints/BackchannelAuthenticationCallbackEndpoint.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.protocol.oidc.grants.ciba.endpoints;
+
+import static org.keycloak.events.Errors.DIFFERENT_USER_AUTHENTICATED;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.jboss.resteasy.util.HttpHeaderNames;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.common.util.Time;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.OAuth2DeviceTokenStoreProvider;
+import org.keycloak.protocol.oidc.grants.ciba.channel.AuthenticationRequest;
+import org.keycloak.protocol.oidc.grants.ciba.channel.HttpAuthenticationChannelProvider;
+import org.keycloak.protocol.oidc.grants.ciba.resolvers.CIBALoginUserResolver;
+import org.keycloak.services.ErrorResponseException;
+
+public class BackchannelAuthenticationCallbackEndpoint extends AbstractCibaEndpoint {
+
+    public static final String SUCCEEDED = "succeeded";
+    public static final String UNAUTHORIZED = "unauthorized";
+    public static final String CANCELLED = "cancelled";
+
+    @Context
+    private HttpRequest httpRequest;
+
+    public BackchannelAuthenticationCallbackEndpoint(KeycloakSession session, EventBuilder event) {
+        super(session, event);
+    }
+
+    @Path("/")
+    @POST
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response processAuthenticationChannelResult() {
```

Comment:
@mposolda @pedroigor +1 . It might be preferable if we could separate "proprietary" part and realize it as an individual class. `AuthenticationChannelProvider` and its implementation classes are ideal for it.

Reply:
Yeah. I suppose that this may not be so big change as we will just need to:
- introduce new method on AuthenticationChannelProvider. Maybe like `handleCallback` or something like that?
- move most of the logic from BackchannelAuthenticationCallbackEndpoint to that new method on AuthenticationChannelProvider . BackchannelAuthenticationCallbackEndpoint might just need to call this AuthenticationChannelPRovider.handleCallback and that's it mostly?

Reply:
@mposolda I agree with your proposal. I would like to change codes as you proposed.

Reply:
@tnorimat Thanks